### ISSUE TITLE
In `jni`, java object's external sizes seems not be passed to Dart VM

### ISSUE THREAD

**Issue by fzyzcjy** [dart-lang/native#849]
Hi thanks for the library! In https://github.com/fzyzcjy/flutter_rust_bridge, it has long been a problem to correctly provide `externalSize` value for the `NativeFinalizer.attach()` function. Because, you know, it is hard to say how much memory an object really takes, because it can point to another object, and the other object can point to a third object, etc.

Thus I want to have a look at how `jnigen`/`jni` does so. Surprisingly, it does nothing!

https://github.com/dart-lang/native/blob/5aeab2915167fd24ccb1241c07516b388fda2b9b/pkgs/jni/lib/src/jreference.dart#L45

It seems that it just ignores this paramter.

Therefore, I wonder 

* Is this allowed, or even suggested behavior?
* Is it better to provide an inaccurate externalSize (e.g. in flutter_rust_bridge, I just provide `sizeof(TheType)` ignoring any outgoing pointers that may also contain memory), or provide null (jni does this)?

**Comment by HosseinYousefi** [comment]
cc/ @dcharkes

**Comment by dcharkes** [comment]
Adding `externalSize` creates extra GC pressure for the Dart GC to GC that object. The idea being that when the Dart object is GCed, the native object can be freed.

In the context of JNIgen, we have have the Java GC. So the only thing that we can free in the native finalizer is the JNI global handle which lives in C memory. Whether the Java object is freed depends on when the Java GC is running. So the only right thing to do here is to add the size of the JNI global handle as external size.
(I presume these handles are fairly small, possibly the same order of magnitude as a Dart object. So leaving out the external size doesn't really influence the behavior.)

For flutter_rust_bridge, where there is no GC, the native finalizer might actually be freeing more. In cases where you have large structs or fixed size arrays, that when freed do not free anything they're pointing to, you can add the correct external size.

Basically, it's a way to make the Dart GC run a bit earlier (because more size in the accounting) so that the native objects are freed earlier.

If you over-estimate the external size, you might drive the GC crazy, as it will try to collecting garbage to free some memory, but the native memory isn't actually released.

If you under-estimate the external size significantly (e.g. you have a 100 MB buffer that you're holding on to from Dart), but report 0 bytes. The Dart GC might not run at all because it expects not much to be gained.

TL;DR:

* Don't follow JNIgen, because it has a Java GC which makes it a completely different scenario.
* Only give the external size for things that are guaranteed to be freed on running the native finalizer. (Don't over estimate.)
* Do give an external size for very large native memory chunks.

@rmacnak-google might be able to give more insights here.

**Comment by fzyzcjy** [comment]
Thank you very much for the suggestion!

Then I guess the best thing I can do is to add an API for the user to specify their actual external size... Because it is almost impossible to know the actual size of a Rust (e.g.) struct. For example, if the struct *owns* another big buffer, we can count it; but if it shares (e.g. using reference counting) a big buffer with other objects, then we cannot count it. Things can even be dynamic (consider a vector with changing length).

**Comment by mkustermann** [comment]
> In the context of JNIgen, we have have the Java GC. So the only thing that we can free in the native finalizer is the JNI global handle which lives in C memory. Whether the Java object is freed depends on when the Java GC is running. So the only right thing to do here is to add the size of the JNI global handle as external size.
> 
> (I presume these handles are fairly small, possibly the same order of magnitude as a Dart object. So leaving out the external size doesn't really influence the behavior.)

That's not really the right reasoning here: Imagine a Dart object is the only thing hanging (via a global JNI handle) onto a Java object. That Java object references 100s of MBs of resources. Now if we don't incentivize the Dart VM's GC to run and collect the small Dart object, we will not delete the global JNI handle. As long as the global JNI handle isn't deleted, the Java GC cannot reclaim those 100 MB.

In the end with native resources (we can somewhat consider the JNI handles as native resources) that can be very large are better handled via "manual memory management", eager disposal.
